Client yaml config

erosb opened the pull request:
Follow-up of https://github.com/hazelcast/hazelcast/pull/18441
with client config schema, and client failover config schema.

Commit:
Invalidate cached SQL plans on mapping changes (#18556)

Commit:
Support dynamic parameters in Jet SQL engine (#18522)

frant-hartm committed:
Update distribution zip contents (#18576)

Based on the discussion we want to have the aws, azure, gcp and kubernetes modules in both distributions.

The spring module is not required on the server side.
(hibernate and wm module are required for certain scenarios).

* Unify the assembly header

Commit:
Merge SQL modules (#18584)

Commit:
Exclude generated sources from checkstyle checks (#18587)

dependabot[bot] committed:
Bump jline.version from 3.16.0 to 3.19.0 (#18434)

Bumps `jline.version` from 3.16.0 to 3.19.0.

Updates `jline-terminal` from 3.16.0 to 3.19.0
- [Release notes](https://github.com/jline/jline3/releases)
- [Changelog](https://github.com/jline/jline3/blob/master/changelog.md)
- [Commits](https://github.com/jline/jline3/compare/jline-parent-3.16.0...jline-parent-3.19.0)

Updates `jline-reader` from 3.16.0 to 3.19.0
- [Release notes](https://github.com/jline/jline3/releases)
- [Changelog](https://github.com/jline/jline3/blob/master/changelog.md)
- [Commits](https://github.com/jline/jline3/compare/jline-parent-3.16.0...jline-parent-3.19.0)

Updates `jline-terminal-jansi` from 3.16.0 to 3.19.0
- [Release notes](https://github.com/jline/jline3/releases)
- [Changelog](https://github.com/jline/jline3/blob/master/changelog.md)
- [Commits](https://github.com/jline/jline3/compare/jline-parent-3.16.0...jline-parent-3.19.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ahmetmircik committed:
Add toString method to IndexImpl (#18561)

Add detailed failure message to assertion. This will help to pinpoint issue better.

Fly-Style committed:
[PLAT-167] SQL : imlement BETWEEN operator support (#18483)

* SQL : basic BETWEEN operator support

* Implement general behaviour test

* Add operand type checker

* Add dates and null check tests

* Minor changes

* Fix an issue with infinite recursion calls in makeLiteral(); add more tests

* Apply review comments

* Fix checkstyle errors, add tests for allowed explicit type casts

* Stabilize tests

* Reorganize tests for BETWEEN

* Introduce BetweenOpNumericPrecedenceParameterConverter

* Small nesting test fix, remove javafx Pair from test

* Simplify ExpressionType

* Rewrite javadoc, use Tuple2 instead of a custom class

* Move inner classes to the end

* Inline methods acting as single-use string constants

* A test with casts didn't actually cast

* Test isn't parametric

* Auto-format code

* Simplify the test, add a failing case

* Test also with OBJECT type

* Implement separate BETWEEN type inferencer

Co-authored-by: Viliam Durina <viliam-durina@users.noreply.github.com>

viliam-durina committed:
Fix Jet client protocol files generator (#18589)

Fly-Style committed:
[PLAT-166] Implement basic IN operator support (#18422)

* SQL: implement basic IN operator support
* SQL: creates hook to SqlToRelConverter to override IN operator behaviour
* SQL: ignore NULL inside IN operator clause
* SQL : rework IN operator inheritance, fix tests
* Rearrange tests, fix null-check tests
* Add dynamic parameter tests
* Extend test collection for IN operator
* Apply suggestions from code review, start to rework type checker
* Support NOT (this IN ...) statements
* Disable stringPromotion in getWiderTypeFor
* Minor changes
* Minor test enhancement
* Apply review notes

Co-authored-by: Alex Dukhno <5074607+alex-dukhno@users.noreply.github.com>
Co-authored-by: Viliam Durina <viliam-durina@users.noreply.github.com>

sancar committed:
Dont remove backup listeners on connection problem

The client backup listeners are kept per endpoint(connection) as
all listeners are.

When an endpoint is gone, the removal of related listeners
done automatically.

In the problematic scenario, even though the client is connected
back, we remove the backup listener anyway. This is because we
use `clientUuid` as the key for backup listener.

I have changed removal so that it takes the `value` into account.
This way, an old scheduled autamatic removal can not delete a more
up-to-date listener from backupListeners map.

fixes https://github.com/hazelcast/hazelcast/issues/18596

viliam-durina committed:
Rename params to arguments or "param values" (#18594)

* Rename params to arguments or "param values"

This is to unify terminology: query has parameters, and for each
parameter we provide an argument or a value

* Remove failing but pointless test

Commit:
Cleanup SQL test output (#18601)

ufukyilmaz committed:
Add null check to client exception factory in JetService initialization (#18600)

ClientExceptionFactory becomes null when ClientEngine is
NoOpClientEngine (ClientEngine can be NoOpClientEngine according to
network config).

vbekiaris committed:
Allow cluster version change (#18585)

to support 4.2 -> 5.0 rolling upgrade.

Commit:
Fix named arguments + dynamic parameters type mismatch handling (#18603)

Commit:
Make file table functions implement HazelcastFunction (#18602)

viliam-durina committed:
Share cluster in ExpressionTestSupport (#18599)

Reduces the affected tests form 69 seconds to about 44 seconds locally.
Also makes the laptop more quiet during that.

alex-dukhno committed:
Add support for CASE operator (#18067)

Commit:
Remove printing the license key while starting instance with overriding config (#18568)

* unmask some part of license key

ufukyilmaz committed:
Update README.md and CONTRIBUTING.md (#18552)

Commit:
reorganizing test structure to support client-only and member-only and common tests

Commit:
adjusting instance pointers in expected error jsons when running client tests

Commit:
[PLAT-200] Implementation of SQL EXTRACT function (#18570)

Adds support for `EXTRACT(field FROM source)` function. 

The function computes date parts from the source field. 
Example:
```
SELECT EXTRACT(DAY FROM TIMESTAMP '2021-04-20 14:35:00');
20
```

Supported types for `source` argument:
- Date
- Time
- Timestamp
- Timestamp With Time Zone

Supported enumeration for `field` argument:
- MILLENNIUM

  The first millennium starts at 0001-01-01 00:00:00, the second century starts at 1001-01-01 00:00:00 etc.

- CENTURY

  The first century starts at 0001-01-01 00:00:00, the second century starts at 0101-01-01 00:00:00 etc.

- DECADE

  Year divided by 10.

- YEAR

  Extracts the year field from the source.

- ISOYEAR

  Extracts the ISO 8601 week-numbering year from the source.

- QUARTER
  
  Extracts the quarter (1-4) of the year from the date source.
   
- MONTH
  
  Extracts the month (1-12) of the year from the date source.

- WEEK

  Extracts the ISO 8601 week-numbering week of the year from the source.

- DOW
  
  Returns the day of the week starting from Sunday (0) to Saturday (6).

- ISODOW

  Returns the day of the week starting from Monday (1) to Sunday (7).

- DAY

  Extract the day part (1-31) from the source.

- HOUR

  Extracts the hour part (0-23) from the source.

- MINUTE

  Extracts the minute part (0-59) from the source.

- SECOND
  
  Extracts the second part from the source without discarding fraction part.

- MILLISECOND
  
  Returns the `SECOND` field multiplied with `1000`.

- MICROSECOND

  Returns the `SECOND` field multiplied with `1000000`.

- EPOCH

  Returns the number of seconds since `1970-01-01 00:00:00+00:00`.


Caveats:

- TIMESTAMP WITH TIME ZONE literals are not supported but it should work correctly
   with the `OffsetDateTime` objects.
- BC date-time literals like `0023-02-20 BC` or `-0023-02-20` are not supported but they 
  should work correctly if the are represented as `java.time` objects.
- INTERVAL type is not supported yet.

Commit:
Support dynamic parameters in file table functions (#18613)

Commit:
Remove client endpoint on connection remove (#18562)

Client endpoints were not destroyed on connection remove but
instead a periodic task was cleaning them up(every 10 seconds)
This was causing some tests to take long.
Added client connections to server connection manager so that
connectionRemoved can be called and related endpoint is
removed as soon as the connection is removed.

While adding the test, I have realized that we were registering
client connection to the server on the mock connection manager.
This behaviour is different than actual one. This was hiding the
bug on mock tests. Removed that registration. The actual
registration is done while authentication.

fixes #18318

Commit:
Fixes a race on bluegreen (#18511)

Commit:
client config schema wip

 * implemented the following client config sections by referring to their member config equivalent: import, config-replacers, native-memory, listeners, serialization
 * moved ^their corresponding tests to the `common/` dir so that they are run by both `ClientYamlSchemaTest` and `MemberYamlSchemaTest`
 * schematized the following client config sections: client-labels, load-balancer, proxy-factories (and the simple scalar props)

ahmetmircik committed:
Control temporary allocations for background expiry task (#18633)

frant-hartm committed:
Fix rebalance() nonfunctional before fused stages (#18586)

* Add reproducer for rebalance in stage which is later fused
* Set the rebalance flag of fused stage

Co-authored-by: Ufuk Yılmaz <ufuk.yilmaz@hazelcast.com>

Commit:
schematizing client near-cache, adding missing near-cache.local-update-policy

Commit:
Support new Portable types for auto-discovered IMaps (#18639)

Commit:
schematizing client flake-id-generator

Commit:
Make the naming for timestamps consistent (#18637)

Commit:
schematized query caches

Commit:
reliable-topic client-side config schema

Commit:
adding user-code-deployment client config schema

Commit:
wiring config schema validation into YamlClientConfigBuilder

Commit:
instance-tracking config schema

Commit:
schematized client metrics config

Commit:
Make EXTRACT function work on multiple types (#18631)

- EXTRACT function implicitly casts its argument to
  TIMESTAMP WITH TIME ZONE to calculate fields. Therefore,
  the results might be wrong if the argument has local
  temporal type (DATE, TIMESTAMP).
- Indirectly, we can conform Postgres behavior by altering
  how field extraction works for TIME type. Postgres does
  not allow extraction of fields like DOW, DAY etc. from
  TIME type. This PR applies Postgres-like behavior.

Co-authored-by: Viliam Durina <viliam-durina@users.noreply.github.com>

Commit:
adding schema for connection strategy

Commit:
schema refactoring: extracting client subschemas to named subschemas under definitions

Commit:
add jmx_prometheus_javaagent jar into slim distribution zip (#18641)

- upgrade jmx_prometheus_javaagent jar to 0.14.0

Commit:
Remove unused TypedOperandCheckers (#18643)

mdumandag committed:
Pretty print cluster connect timeout in logs (#18642)

The default value for the cluster connect timeout is infinite,
so it is represented with `Long.MAX_VALUE`. When we try to log
this as it is, we output something like below.

```
... cluster connect timeout: 9223372036854775807 ms, ...
```

To make this more clear, we will replace this with the
output below.

```
... cluster connect timeout: INFINITE, ...
```

If the user has provided some other timeout value (such as 999 ms),
the output will still be as it is now.

```
... cluster connect timeout: 999 ms, ...
```

Commit:
schematized client security config

mdumandag committed:
Fix handling of predicates for ReplicatedMap (#18623)

ReplicatedMap was passing null extractors while constructing the QueryEntry
which was cauising NullPointerExceptions while evaluating predicates
that has an attribute path.

Also, refactored the ReplicatedMapListenerTests to reduce test duplication.
Note that, test logic is not altered.

Commit:
schematized client network config

Commit:
failure testcase for client network config

Commit:
adding root node validation as programmatic check

frant-hartm committed:
Fix bootstrappedInstance() when run directly (#18653)


Fixes #18627

Commit:
merging projectx-yaml-schema , resolving conflicts

Commit:
checkstyle fixes

Commit:
client-failover config schema and some test fixes

 * added schema validation step to `XmlClientConfigBuilder`, `XmlClientFailoverConfigBuilder`, `YamlClientFailoverConfigBuilder`
 * added `hazelcast-client-failover` as permitted config root element to `YamlConfigSchemaValidator`
 * fixed description in `hazelcast-client-failover-full-example.yaml` (previously it was copy-pasted from the xml)
 * schematized `hazelcast-client-failover` root node
 * removed `YamlClientConfigImportVariableReplacementTest#testImportNoHazelcastClientRootNode()` since we don't support configs without proper root node anymore
 * now if a failover config defines 0 client config files, we throw a `SchemaViolationConfigurationException` instead of `InvalidConfigurationException` (test updated accordingly)
 * removed non-existent `connection-retry.enabled=false` property from `hazelcast-client-c1.yaml`
 * fixed schema violation in `hazelcast-client-c2.yaml`

viliam-durina committed:
Fix MetricsTest

I think due to some test setup the `localParallelism !=
availableProcessors`. This PR makes the test not depend on that
equality.

Fixes #18626

Commit:
Fixed remaining tests, schematizing kerberos config

 * renamed `backup-ack-to-client` to `backup-ack-to-client-enabled`, since this is the correct (parsed) property name
 * added missing kerberos identity config
 * fixed test failures

Commit:
checkstyle fixes

mdumandag committed:
Wait until client receives member shutdown in ClientCacheThroughHazelcastInstanceTest (#18658)

In `getCache_whenOwnerInstanceIsShutdown_thenOperateOnCacheFails`, we shutdown the owner
instance, and than make a put request from the client. On a sunny day, we expect to
get HazelcastClientNotActiveException since the request is made after calling
the instance shutdown.

But on some occasions, it might happen that although the member shutdown process
is complete, the client might not be aware of it yet. If we make a put request
in this scenario, the client will try to route put request to shutdown member,
get the connection close event some time after, set the invocation with
the TargetDisconnectedError. Since the put request is not retryable, the client
will finalize the put request with this exception and the test will fail.

As a solution, before making the put request, we wait until the connection
related to the shutdown member is removed from the client.

jerrinot committed:
Always set `dbConnectionUrl` as a first thing of the test class initialization (#18659)

Reasoning:
Previously an exception in the `initialize(2, null);` meant `dbConnectionUrl`
was left a NULL. But `afterClass()` does not expect `dbConnectionUrl` to be null
and `DriverManager.getConnection()` caught an exception obfustacting the original
root cause: a failure in `initialize(2, null);`

Commit:
Correct SQL type sizes (#18662)

Commit:
Limit number of records that can be accumulated by Jet processors (#18671)

Commit:
Add SQL grouping memory management test (#18679)

Commit:
Replace MetaSupplierCtx.processingGuarantee with a call to JobConfig.getProcessingGuarantee() (#18674)

Commit:
Apply maxProcessorAccumulatedRecords to TransformStatefulP (#18678)

jbartok committed:
Cleanly skip split brain protection testing for Jet operations (#18617)

Commit:
Add support for job-level maxProcessorAccumulatedRecords (#18677)

Commit:
Apply maxProcessorAccumulatedRecords to DistinctTransform (#18682)

Commit:
Add reverse javadocs to memory limited operations (#18683)

Commit:
Add memory management design doc (#18684)

lprimak committed:
Remove sleep statements - Reduce delays when joining a cluster (#18267)

Eliminated delays when joining a cluster. Removed sleep statements and replaced them with events

frant-hartm committed:
Fix default cluster name for hazelcast command (#18687)

blazember commented on hazelcast/src/main/resources/hazelcast-client-full-example.yaml:
```diff
@@ -231,7 +231,7 @@ hazelcast-client:
   #         your system’s default TTL.
   #     * "echo-fail-fast-on-startup":
-  #         Cluster Member will fail to start if it is unable to action an ICMP ping command when ICMP is enabled.
+  #         The client will fail to start if it is unable to action an ICMP ping command when ICMP is enabled.
```
:+1:

blazember commented on hazelcast/src/test/java/com/hazelcast/client/config/XmlClientFailoverConfigBuilderTest.java:
```diff
@@ -128,7 +128,7 @@ ClientFailoverConfig buildConfig() {
     }
 
-    private static ClientFailoverConfig buildConfig(String yaml) {
+    public static ClientFailoverConfig buildConfig(String yaml) {
```
Why these became public?

erosb replied:
They are used by `SchemaValidationSkippingTest` as well. I didn't want to duplicate the constants and the `buildConfig()` implementation so I made them public instead.

blazember commented on hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigImportVariableReplacementTest.java:
```diff
@@ -81,24 +82,6 @@ public void testImportElementOnlyAppearsInTopLevel() throws IOException {
         assertNull(clientConfig.getInstanceName());
```
I think this assertion is a dead code then and can be removed.

blazember reviewed (approved):
Went through only the java files. In general, it looks nice, but the PR has to be approved in the upstream as well once again, so I'll need to do a final check and give my final thumbs up there in a separate PR.

viliam-durina committed:
Light Jet jobs (#18619)

See `019-light-jobs.md` included in the PR.

Fixes #18605.

Commit:
Replace usage of hasMessageEndingWith() with hasMessageContaining() in SQL tests (#18689)

kwart committed:
Add OWASP dependency checker configuration (#18563)

Commit:
removing unused assert

Commit:
Merge branch 'projectx-yaml-validation' into client-yaml-config

Commit:
YAML config validation (#18441)

PRD: https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/3106963474/YAML+configuration+validation

## Summary of code changes

 * adds the previously written json schema for config files to the repo
 * adds a validation step to `YamlConfigBuilder#parseAndBuildConfig()`, this throws a `SchemaViolationConfigurationException` (which extends `InvalidConfigurationException`)
 * adds [this library](https://github.com/everit-org/json-schema) as a shaded maven dependency which performs the actual validation of the yaml config against the schema. Note: this is an implementation-specific detail, so the library can be swapped out without breaking the API
 * `YamlToJsonConverter` before calling the validator, this class is used to convert the config from its snakeyaml representation to org.json representation
 * there are also plenty of declarative tests as json files, which test the schema file itself with valid and invalid configurations. *There is no need to review these one by one* since the schema file changes (and tests) was previously reviewed in detail in [this repository](https://github.com/erosb/hz-yaml-config-schema).

## Breaking changes

 * case insensitive enum values are not supported in config anymore
 * the top-level `hazelcast` property is now mandatory (previously it was optional, but it was an undocumented feature)
 * small chance: if someone mixed HZ YAML config with an other YAML-based config,  then previously they could start the member with that mixed config file. Now it won't start up because the unrecognized properties will fail. There is a system property added to work this around (`-Dyaml.config.validation.skip`).

Co-authored-by: Pavlo Romanenko <pavlo.romanenko@hazelcast.com>
Co-authored-by: Iurii Kutelmakh <yuraku@users.noreply.github.com>

Commit:
merging master, resolving conflicts

Commit:
fixing compilation failures and removing duplicate test json directories

Commit:
deleting YamlSchemaTest

These tests are now run by AbstractYamlSchemaTest and its subclasses. The YamlSchemaTest class was a leftover from merging master
